<h1><span style="color: lightblue;">Performing Data Wrangling ⚡️</span></h1>

## 🌟 Step 1
<h2><span style="color: yellow;">Data Gathering 🌎 </span></h2> 

   From various sources :<br>
- `CSV files` <br>

In [2]:
#Importing nessesary modules
import pandas as pd
import numpy as np

In [3]:
# Reading CSV
titanic_data = pd.read_csv('Titanic-Dataset.csv')
titanic_data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

## 🌟 Step 2
<h2><span style="color: yellow;">Data  Assessments 🌎 </span></h2>  
- In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.<br>
- Basically a whole summary of data.<br>
- Data assessment is often an iterative process.<br>
    
<h2><span style="color: red;"> 🛑 Step 1 Discover </span></h2>  
- View datasets.<br>
- Check shape of the data.<br>
    
### 🔥 Automatic Assessments 
- Programmatic.<br>
- Using Pandas.<br>
  - `head and tail`
  - `describe`
  - `sample`
  - `info`
  - `isnull`
  - `duplicated`

In [34]:
# Display random 50 sample
titanic_data.sample(50)

PassengerId  Survived  Pclass  \
344          345         0       2   
516          517         1       2   
211          212         1       2   
883          884         0       2   
28            29         1       3   
403          404         0       3   
766          767         0       1   
430          431         1       1   
115          116         0       3   
53            54         1       2   
551          552         0       2   
401          402         0       3   
325          326         1       1   
10            11         1       3   
456          457         0       1   
88            89         1       1   
7              8         0       3   
851          852         0       3   
789          790         0       1   
602          603         0       1   
286          287         1       3   
689          690         1       1   
765          766         1       1   
683          684         0       3   
475          476         0       1   
207          208         1       3   
616          617         0       3   
623          624         0       3   
629          630         0       3   
273          274         0       1   
658          659         0       2   
263          264         0       1   
123          124         1       2   
644          645         1       3   
445          446         1       1   
291          292         1       1   
365          366         0       3   
831          832         1       2   
313          314         0       3   
439          440         0       2   
482          483         0       3   
815          816         0       1   
596          597         1       2   
874          875         1       2   
359          360         1       3   
724          725         1       1   
54            55         0       1   
555          556         0       1   
170          171         0       1   
546          547         1       2   

                                                  Name     Sex    Age  SibSp  \
344                            Fox, Mr. Stanley Hubert    male  36.00      0   
516                       Lemore, Mrs. (Amelia Milley)  female  34.00      0   
211                         Cameron, Miss. Clear Annie  female  35.00      0   
883                      Banfield, Mr. Frederick James    male  28.00      0   
28                       O'Dwyer, Miss. Ellen "Nellie"  female    NaN      0   
403                     Hakkarainen, Mr. Pekka Pietari    male  28.00      1   
766                          Brewe, Dr. Arthur Jackson    male    NaN      0   
430          Bjornstrom-Steffansson, Mr. Mauritz Hakan    male  28.00      0   
115                              Pekoniemi, Mr. Edvard    male  21.00      0   
53   Faunthorpe, Mrs. Lizzie (Elizabeth Anne Wilkin...  female  29.00      1   
551                        Sharp, Mr. Percival James R    male  27.00      0   
401                                    Adams, Mr. John    male  26.00      0   
325                           Young, Miss. Marie Grice  female  36.00      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.00      1   
456                          Millet, Mr. Francis Davis    male  65.00      0   
88                          Fortune, Miss. Mabel Helen  female  23.00      3   
7                       Palsson, Master. Gosta Leonard    male   2.00      3   
851                                Svensson, Mr. Johan    male  74.00      0   
789                           Guggenheim, Mr. Benjamin    male  46.00      0   
602                          Harrington, Mr. Charles H    male    NaN      0   
286                            de Mulder, Mr. Theodore    male  30.00      0   
689                  Madill, Miss. Georgette Alexandra  female  15.00      0   
765               Hogeboom, Mrs. John C (Anna Andrews)  female  51.00      1   
683                        Goodwin, Mr. Charles Edward    male  14.00      5   
475                        Clifford, Mr. George Quincy    male    

In [9]:
# Count Survived people in the titanic data
titanic_data[titanic_data['Survived'] == 1].count()

PassengerId    342
Survived       342
Pclass         342
Name           342
Sex            342
Age            290
SibSp          342
Parch          342
Ticket         342
Fare           342
Cabin          136
Embarked       340
dtype: int64

In [8]:
# Sum of duplicated colomn in titanic_data
num_duplicates = titanic_data.duplicated().sum()
num_duplicates

0

In [7]:
# Expose null values in a  dataset
titanic_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
# Statistcal detail of all numerical colomn in dataset
titanic_data.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

### 🔥 Manual Assessments 
- Export data into excel sit and watch for hours to the data or looking through the data manually in google sheets.

In [6]:
# Exporting data into excel sheet
with pd.ExcelWriter('titanic_dataset.xlsx') as writer:
  titanic_data.to_excel(writer,sheet_name='titanic_data')

<h2><span style="color: red;"> 🛑 Step 2 Document </span></h2> 
- Summary (**refer to readme.md**) <br>
- Address issues within the dataset combine and make documents<br>

- **The Titanic Tragedy Dataset**
- The RMS Titanic set sail on its maiden voyage in 1912, only to become infamous for its tragic sinking. This dataset provides insights into the passengers aboard the ill-fated ship.
- **Dataset Overview**: Contains information about 891 passengers, including their survival status.
- **Survival Statistics**: Of the 891 passengers, 549 did not survive, while 342 survived the disaster.
- **Passenger Information**: The dataset includes various details about the passengers, such as their age, class, and ticket information, providing a glimpse into the demographics and circumstances of those on board.

## 🌟 Step 3
<h2><span style="color: yellow;">Data Cleaning or Data Quality Dimensions✨</span></h2> 

In [29]:
# Making a copy of dataset for analysis
titanic_data_df = titanic_data.copy()
titanic_data_df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [30]:
# Fill missing Age values with median
titanic_data_df['Age'].fillna(titanic_data_df['Age'].median(), inplace=True)

# Fill missing Embarked values with the most common value
titanic_data_df['Embarked'].fillna(titanic_data_df['Embarked'].mode()[0], inplace=True)

# Indicate missing Cabin values as 'Unknown'
titanic_data_df['Cabin'].fillna('Unknown', inplace=True)

In [31]:
# Drop duplicates
titanic_data_df.drop_duplicates(inplace=True)

In [32]:
# Change data type 
titanic_data_df['Age'] = titanic_data_df['Age'].astype(int)

In [33]:
# Drop useless colomns for analysis
titanic_data_df.drop(['PassengerId', 'Ticket', 'Cabin'], axis=1, inplace=True)

## 🌟 Step 4
<h2><span style="color: yellow;">Feature Engineering ✨</span></h2> 

In [34]:
import re

def separate_name(full_name):
    # Regular expression to capture the three parts
    pattern = r'^([^,]+), ([^(]+)(?:\(([^)]+)\))?$'
    match = re.match(pattern, full_name)
    
    if match:
        name = match.group(1).strip()
        sure_name = match.group(2).strip()
        nickname = match.group(3).strip() if match.group(3) else None
        return (name, sure_name, nickname)
    else:
        return (None, None, None)

In [35]:
# Making name colomn seperate for name ,surename ,nickname
titanic_data_df['Name_1'] = titanic_data_df['Name'].apply(lambda x: separate_name(x)).apply(lambda x:x[0])
titanic_data_df['Sure Name'] = titanic_data_df['Name'].apply(lambda x: separate_name(x)).apply(lambda x:x[1])
titanic_data_df['Nick Name'] = titanic_data_df['Name'].apply(lambda x: separate_name(x)).apply(lambda x:x[2])

In [36]:
# Drop Name colomn
titanic_data_df.drop(columns='Name',inplace=True)

In [37]:
titanic_data_df['Name'] = titanic_data_df['Name_1']

In [38]:
# Drop Name colomn
titanic_data_df.drop(columns='Name_1',inplace=True)

In [85]:
# Outliers 0 fare
titanic_data_df[titanic_data_df['Fare'] == 0]

PassengerId  Survived  Pclass   Sex  Age  SibSp  Parch  Ticket  Fare  \
179          180         0       3  male   36      0      0    LINE   0.0   
263          264         0       1  male   40      0      0  112059   0.0   
271          272         1       3  male   25      0      0    LINE   0.0   
277          278         0       2  male   28      0      0  239853   0.0   
302          303         0       3  male   19      0      0    LINE   0.0   
413          414         0       2  male   28      0      0  239853   0.0   
466          467         0       2  male   28      0      0  239853   0.0   
481          482         0       2  male   28      0      0  239854   0.0   
597          598         0       3  male   49      0      0    LINE   0.0   
633          634         0       1  male   28      0      0  112052   0.0   
674          675         0       2  male   28      0      0  239856   0.0   
732          733         0       2  male   28      0      0  239855   0.0   
806          807         0       1  male   39      0      0  112050   0.0   
815          816         0       1  male   28      0      0  112058   0.0   
822          823         0       1  male   38      0      0   19972   0.0   

       Cabin Embarked      Name_1                  Sure Name Nick Name  
179  Unknown        S     Leonard                 Mr. Lionel      None  
263      B94        S    Harrison                Mr. William      None  
271  Unknown        S   Tornquist          Mr. William Henry      None  
277  Unknown        S      Parkes        Mr. Francis "Frank"      None  
302  Unknown        S     Johnson     Mr. William Cahoone Jr      None  
413  Unknown        S  Cunningham         Mr. Alfred Fleming      None  
466  Unknown        S    Campbell                Mr. William      None  
481  Unknown        S       Frost  Mr. Anthony Wood "Archie"      None  
597  Unknown        S     Johnson                 Mr. Alfred      None  
633  Unknown        S        Parr    Mr. William Henry Marsh      None  
674  Unknown        S      Watson         Mr. Ennis Hastings      None  
732  Unknown        S      Knight               Mr. Robert J      None  
806      A36        S     Andrews              Mr. Thomas Jr      None  
815     B102        S         Fry                Mr. Richard      None  
822  Unknown        S    Reuchlin      Jonkheer. John George      None

In [40]:
# Dealing outliers from median of fare coloumn
median_fare = titanic_data_df['Fare'].median()
titanic_data_df.loc[titanic_data_df['Fare'] == 0, 'Fare'] = median_fare

- Create New Features:
- `Family Size` : Combine SibSp and Parch to create a new feature FamilySize.

In [41]:
# New feature family size
titanic_data_df['FamilySize'] = titanic_data_df['SibSp'] + titanic_data_df['Parch'] + 1

IsAlone: Create a feature to indicate if a passenger is alone.

In [95]:
# New feature IsAlone  
titanic_data_df['IsAlone'] = 1
# Making 0 in IsAlone if familysize greater than 1
titanic_data_df.loc[titanic_data_df['FamilySize'] > 1, 'IsAlone'] = 0

In [42]:
# The pd.cut() function is used to segment the 'Age' column into bins (age intervals)
# bins specifies the edges of the intervals: [0, 12, 20, 40, 60, 80]
# labels specifies the names for each bin: ['Child', 'Teen', 'Adult', 'Middle-Aged', 'Senior']
titanic_data_df['AgeGroup'] = pd.cut(
                                    titanic_data_df['Age'], 
                                    bins = [0, 12, 20, 40, 60, 80], 
                                    labels = ['Child', 'Teen', 'Adult', 'Middle-Aged', 'Senior'])

`pd.cut() in Pandas is named for its primary functionality: to "cut" a continuous variable into discrete intervals or bins.`

In [45]:
titanic_data_df.isnull().sum()

Survived        0
Pclass          0
Sex             0
Age             0
SibSp           0
Parch           0
Fare            0
Embarked        0
Sure Name       2
Nick Name     750
Name            2
FamilySize      0
AgeGroup        7
dtype: int64

In [53]:
titanic_data_df.drop(columns = 'Nick Name',inplace=True)

In [46]:
def family_size_group(size):
    if size == 1:
        return 'Single'
    elif size <= 3:
        return 'Small Family'
    else:
        return 'Large Family'

In [48]:
# New feature FamilySizeGroup
titanic_data_df['FamilySizeGroup'] = titanic_data_df['FamilySize'].apply(family_size_group)

In [56]:
# Create new features based on interactions between existing features, such as 'Fare per Family Member'.
titanic_data_df['FarePerFamilyMember'] = titanic_data_df['Fare'] / titanic_data_df['FamilySize']

In [58]:
# Exporting to CSV
titanic_data_df.to_csv('processed_titanic_dataset.csv', index=False)

In [57]:
titanic_data_df

Survived  Pclass     Sex  Age  SibSp  Parch     Fare Embarked  \
0           0       3    male   22      1      0   7.2500        S   
1           1       1  female   38      1      0  71.2833        C   
2           1       3  female   26      0      0   7.9250        S   
3           1       1  female   35      1      0  53.1000        S   
4           0       3    male   35      0      0   8.0500        S   
..        ...     ...     ...  ...    ...    ...      ...      ...   
886         0       2    male   27      0      0  13.0000        S   
887         1       1  female   19      0      0  30.0000        S   
888         0       3  female   28      1      2  23.4500        S   
889         1       1    male   26      0      0  30.0000        C   
890         0       3    male   32      0      0   7.7500        Q   

                          Sure Name       Name  FamilySize AgeGroup  \
0                   Mr. Owen Harris     Braund           2    Adult   
1                 Mrs. John Bradley    Cumings           2    Adult   
2                       Miss. Laina  Heikkinen           1    Adult   
3                Mrs. Jacques Heath   Futrelle           2    Adult   
4                 Mr. William Henry      Allen           1    Adult   
..                              ...        ...         ...      ...   
886                     Rev. Juozas   Montvila           1    Adult   
887            Miss. Margaret Edith     Graham           1     Teen   
888  Miss. Catherine Helen "Carrie"   Johnston           4    Adult   
889                 Mr. Karl Howell       Behr           1    Adult   
890                     Mr. Patrick     Dooley           1    Adult   

    FamilySizeGroup  FarePerFamilyMember  
0      Small Family              3.62500  
1      Small Family             35.64165  
2            Single              7.92500  
3      Small Family             26.55000  
4            Single              8.05000  
..              ...                  ...  
886          Single             13.00000  
887          Single             30.00000  
888    Large Family              5.86250  
889          Single             30.00000  
890          Single              7.75000  

[891 rows x 14 columns]